### PR TITLE
feat(ci): Add codespell to Makefile and GitHub Actions

### PR DESCRIPTION
**Pull request by CharlieTLe**
This is a followup from
https://github.com/open-telemetry/opentelemetry-go/pull/3980 to add the automation for checking for code spelling errors to a different PR.

[Codespell][] makes use of the `.codespellrc` file automatically as a config file for convenience. The configuration file specifies directories and flags to pass to `codespell`. Codespell also makes use of the `.codespellignore` file to ignore certain word spellings.

There is a new section in `Makefile` for installing python tooling. It is similar to how the go tools are installed. It sets up a virtualenv in `venv` and installs python tooling there if it hasn't been installed yet.

The new [codespell workflow][] will run in pull requests to annotate misspelled words. It is set up to only warn the user with annotations in the pull request if there are typos, but we can fail the pull request as well if we want to by deleting the [only_warn][] key.

[codespell]: https://github.com/codespell-project/codespell
[codespell workflow]: https://github.com/codespell-project/actions-codespell
[only_warn]: https://github.com/codespell-project/actions-codespell#parameter-only_warn

cc: @pellared @MrAlias @dmathieu @alolita

### CODE REVIEW

File: .github/workflows/codespell.yaml
Comment (MrAlias) [inline .github/workflows/codespell.yaml:15]: Will this not use the `.codespellrc` that is included in the repository?

File: Makefile
Comment (pellared) [reply Makefile:86]: Maybe using a Docker container?

For reference [here](https://github.com/streetsidesoftware/cspell-cli/pull/183) is what I did for [`cspell`](https://github.com/streetsidesoftware/cspell) so that it could be adopted for non-JS repositories.

File: .github/workflows/codespell.yaml
Comment (CharlieTLe) [reply .github/workflows/codespell.yaml:15]: Inconveniently, no. Although, I do believe the maintainer has a [PR](https://github.com/codespell-project/actions-codespell/pull/19) that should support this in a future version of the action.

File: Makefile
Comment (CharlieTLe) [reply Makefile:86]: I agree. I will work on a new version using a docker version. Thanks for the reviews @MrAlias and @pellared.

**Comment by CharlieTLe** [comment]
Hi @pellared @MrAlias @dmathieu @alolita,

I've updated the PR to use docker to run codespell. Please help review the PR when you get a chance. 

Thank you.

File: .github/workflows/codespell.yaml
Comment (dmathieu) [inline .github/workflows/codespell.yaml:3]:
```suggestion
on:
  push:
    branches:
      - main
  pull_request:
```

**Review by dmathieu — APPROVED** [review]

File: Makefile
Comment (CharlieTLe) [reply Makefile]: Updated PR which should have resolved this conversation.

**Comment by CharlieTLe** [comment]
Thanks for approving @dmathieu! Since I don't have write access, could someone help merge the PR?

**Review by hanyuancheung — APPROVED** [review]

File: Makefile
Comment (pellared) [inline Makefile:107]: Do we want to pin the version of `codespell` or are we OK always using the latest?

The downsides of using the latest would be:
1. The build may suddenly start failing and interrupt other work
2. No control on what is exactly installed (so we have less control if `codespell` somehow does something harmful)

Same comment applies to the GitHub Action.

Maybe we could use `requirements.txt` and control the version there?
Also we could use the Make target instead of the GitHub Action as I do not see if it adds the possibility to configure `codespell` version. 

PS. It may be good to wait for @open-telemetry/go-maintainers for additional feedback if they agree with my comments or not.

Comment (CharlieTLe) [reply Makefile:107]: I do not think we are currently pinning any of the Go tools installed by Make, e.g. misspell, today, but I'm not opposed to pinning codespell to the version that it is currently installing now if that is what is being requested.

> The build may suddenly start failing and interrupt other work (my biggest concern)

The `codespell` target should not be failing any builds since it is not linked as a dependency in the `precommit` target like `misspell` is.

> Maybe we could use requirements.txt and control the version there?
Also we could use the Make target instead of the GitHub Action as I do not see if it adds the possibility to configure codespell version.

We should be able to pin the codespell version like so without the need for a requirements file:

```
(CODESPELL): PACKAGE=codespell==2.2.4
```

Comment (pellared) [reply Makefile:107]: All Go tools are pinned and updated via dependabot thanks to https://github.com/open-telemetry/opentelemetry-go/blob/main/internal/tools/go.mod

> We should be able to pin the codespell version like so without the need for a requirements file:

The downside of it is it will require manual work for updating the dependency.

Comment (CharlieTLe) [reply Makefile:107]: > All Go tools are pinned and updated via dependabot thanks to https://github.com/open-telemetry/opentelemetry-go/blob/main/internal/tools/go.mod

Ah, this is good to know. Thanks for the link.

> The downside of it is it will require manual work for updating the dependency.

Perhaps we can pin it for now and if the issue of upgrading the python dependencies becomes such a chore we can dedicate some engineering effort into automating it?

Comment (pellared) [reply Makefile:107]: > Perhaps we can pin it for now and if the issue of upgrading the python dependencies becomes such a chore we can dedicate some engineering effort into automating it?

I strongly prefer to integrate with dependabot as part of this PR. Otherwise nobody will upgrade the version.

Comment (CharlieTLe) [reply Makefile:107]: So are you recommending that I:

1. Update this PR to create a `requirements.txt` file and pin the version of codespell==2.2.4
2. Create another PR to update https://github.com/open-telemetry/opentelemetry-go-build-tools/tree/main/dbotconf to set up the dependabot configuration for pip
3. Create another PR to with the updated dbotconf to generate the `.github/dependabot.yml` file

?

**Comment by CharlieTLe** [comment]
Hi @pellared, I updated the PR to pin codespell in the requirements.txt file. I'm working on updating dbotconf to support pip (pending internal review) and will submit another PR to generate the new dependabot config file.

In the meantime, does this PR satisfy the first item in the list of recommendations for merging this PR?

File: .github/workflows/codespell.yaml
Comment (pellared) [inline .github/workflows/codespell.yaml:24]: How about 
```suggestion
        run: make codespell
```
?

1. Less dependencies (we would not depend on `codespell-project/actions-codespell` action
2. No need to sync the `make codespell` and `codespell` GitHub workflow

File: .github/workflows/codespell.yaml
Comment (pellared) [inline .github/workflows/codespell.yaml:14]: I do not think we need to fetch all history for all branches and tags

```suggestion
        with:
          fetch-depth: 0
```

File: Makefile
Comment (pellared) [inline Makefile:92]: We could handle docker image updates using dependabot.

For reference:
- https://github.com/signalfx/splunk-otel-go/blob/main/build/markdownlint-cli.dockerfile
- https://github.com/signalfx/splunk-otel-go/pull/1673

_This can be addressed in a separate PR_

**Comment by CharlieTLe** [comment]
@pellared Updated the GH action to use `make codespell`

**Review by pellared — APPROVED** [review]

File: Makefile
Comment (CharlieTLe) [reply Makefile:92]: Thanks @pellared, I will address this in a separate PR.

**Comment by CharlieTLe** [comment]
Thanks for reviewing @pellared.

Can anyone with the 🪄 help merge this PR?

File: .codespellignore
Comment (MrAlias) [inline .codespellignore:4]: Why ignore "collison"? This looks like a misspelling of collision.

Comment (CharlieTLe) [reply .codespellignore:4]: Hi @MrAlias, this is a fair question. There's a person named "Patrick Collison" who probably wouldn't like to be called "Patrick Collision".

https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/jaeger/internal/third_party/thrift/LICENSE#L303

File: .codespellignore
Comment (CharlieTLe) [reply .codespellignore:3]: Yes, for example, `te` is sometimes used as a variable for a TException. For a specific example, please see https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/jaeger/internal/third_party/thrift/lib/go/thrift/exception.go#L37

**Review by MrAlias — APPROVED** [review]

**Comment by CharlieTLe** [comment]
Thanks @MrAlias @pellared

**Pull request merged**
